### DISCUSSION TITLE
Getting Target Property Custom Attributes during OverridePropertySetter?

### DISCUSSION THREAD

**Discussion by bradlee-s** [metalama/Metalama#23] [Q&A] [answered]
This code obviously doesn't work because meta.Target.Property doesn't contain a method GetCustomAttributes like propertyinfo does. But is there a way to get those attributes? If the attribute exists, I'd like to use the LogicalName value as below. Otherwise, proceed with a standard setter. SetAttributeValue is also a method on the class (Xrm.Entity).

Additionally, could I extend my TypeAspect attribute with a LogicalName property that would also decorate the target property with [AttributeLogicalName(LogicalName)]? Then I would know it always exists, and could grab the value easier in my OverridePropertySetter.

		[Template]
		private dynamic OverridePropertySetter(dynamic value)
		{
			if (value != meta.Target.Property.Value)
			{
				try
				{
					var logicalname = meta.Target.Property
						.GetCustomAttribute<AttributeLogicalNameAttribute>.LogicalName;

					this.OnPropertyChanged(logicalname);
					this.SetAttributeValue(logicalname, value);

				}
				catch { meta.Proceed(); }
			}

			return value;
		}

**Comment by gfraiteur** [comment, marked as answer]
To reflect attributes, you can use the [Attributes](https://doc.metalama.net/api/metalama_framework_code_ideclaration_attributes) collection on `meta.Target.Property`.

To add an attribute:

1. Make sure you have the latest build of Metalama - this is a new feature.
2. Override `BuildAspect`.
3. Call `builder.Advice.AddAttribute`. However the attribute added in `BuildAspect` will not be visible in your template method (for the sake of consistency and ease of programming, we consistently expose, to any aspect A, the code model in the state _before_ that aspect A was applied.

You may consider [passing a parameter](https://doc.metalama.net/aspects/templates/template-parameters) from `BuildAspect` to your template.

**Reply by bradlee-s** [reply]
I've installed the newest available version (0.5.25.0), but I don't see Advice.AddAttribute(). I do see .IntroduceAttribute(). Do you have a sample on how to add an attribute to a class and/or property?

**Reply by prochan2** [reply]
This API got renamed in 0.5.25, indeed.

Here is an example:

```cs
using System;
using Metalama.Framework.Aspects;
using Metalama.Framework.Code;
using Metalama.Framework.Code.DeclarationBuilders;

[Inherited] // Apply this attribute if you want your aspect to be inherited.
public class MyAspect : TypeAspect
{
    public override void BuildAspect(IAspectBuilder<INamedType> builder)
    {
        var myAttributeConstruction = AttributeConstruction.Create(typeof(MyAttribute));

        // Apply on the aspect's target type.
        builder.Advice.IntroduceAttribute(builder.Target, myAttributeConstruction, OverrideStrategy.Ignore); // Set the override strategy as required.

        // Apply on an arbitrary type.
        var targetType = TypeFactory.GetType(typeof(TargetClass)?.FullName ?? throw new InvalidOperationException("type not found"));
        builder.Advice.IntroduceAttribute(targetType, myAttributeConstruction, OverrideStrategy.Ignore); // Set the override strategy as required.
    }
}
```

**Reply by bradlee-s** [reply]
>         var myAttributeConstruction = AttributeConstruction.Create(typeof(MyAttribute));
> 
>         // Apply on the aspect's target type.
>         builder.Advice.IntroduceAttribute(builder.Target, myAttributeConstruction, OverrideStrategy.Ignore);

Thanks. This is working! Can you provide an example of how to add a parameter value with the attribute?
[MyAttribute("somevalue")]
I thought something like this might work, but it doesn't.

```
	var myAttributeConstruction = AttributeConstruction.Create(
		typeof(MyAttribute),
		new List<object> { "somevalue" });
	builder.Advice.IntroduceAttribute(builder.Target,
		myAttributeConstruction,
		OverrideStrategy.Ignore);
```

**Reply by addabis** [reply]
@bradlee-s Are you getting a compilation error? I have noticed there is a bug and this case was not tested. It seems only named parameters are now working.

**Reply by bradlee-s** [reply]
@addabis 
I'm unable to see the transformation in VS due to another issue with the extension, so I've been using the interactive online tool. This is the error I'm seeing there when I click Run:
"'MetaEntityAttribute.BuildAspect' threw 'IndexOutOfRangeException' when applied to 'TestEntity': Index was outside the bounds of the array. Exception details are in 'C:\local\Temp\Metalama\CrashReports\0.5.24.0-8df7b802\exception-4086d354-6c52-4d2c-830e-87fda70eea14.log'. To attach a debugger to the compiler, use the  '-p:MetalamaDebugCompiler=True' command-line option.
"

Here's a stripped down version of my code:

`
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

using Metalama.Framework.Aspects;
using Metalama.Framework.Advising;
using Metalama.Framework.Code;
using Metalama.Framework.Code.DeclarationBuilders;

namespace MLTest
{
	[MetaEntity("testing")]
	public class TestEntity
	{
		public Guid Id { get; set; }
		public string Name { get; set; }
	}

	public class TestAttribAttribute : Attribute
	{
		public string Value1 { get; set; }

		public TestAttribAttribute(string value1)
		{
			Value1 = value1;
		}
	}

	[Inherited]
	[AttributeUsage(AttributeTargets.Class, AllowMultiple = false)]
	internal class MetaEntityAttribute : TypeAspect
	{
		private MetaEntityAttribute() { }
		public MetaEntityAttribute(string logicalname = null)
		{
			LogicalName = logicalname;
		}
		public string? LogicalName { get; set; }

		public override void BuildAspect(IAspectBuilder<INamedType> builder)
		{

			var myAttributeConstruction = AttributeConstruction.Create(
				typeof(TestAttribAttribute),
				new List<object> { LogicalName });

			builder.Advice.IntroduceAttribute(builder.Target,
				myAttributeConstruction, OverrideStrategy.Ignore);

		}
	}

}

`

**Comment by addabis** [comment]
@bradlee-s This is the same problem I have encountered, I have fixed it and it should be included in the next release.

**Reply by aweneer** [reply]
Hi @bradlee-s, this has been fixed in the current newest Metalama release 0.5.26-preview.

**Reply by bradlee-s** [reply]
@aweneer I've updated my references from Nuget. Also uninstalled the extension and reinstalled from marketplace (0.5.27). When VS starts up, it displays a warning at the top:"Metalama ran into an unexpected problem: Cannot find the Metalama.Compiler.Interface assembly."
Details:
Metalama.Framework.Engine.AssertionFailedException: Cannot find the Metalama.Compiler.Interface assembly.
   at Obfuscated.^5Nx5YQeXj2J2..ctor(IServiceProvider serviceProvider)
   at Metalama.Framework.Engine.Pipeline.ServiceProviderFactory.<>c.^0Nzc6IUSPbPz(IServiceProvider _0)
   at Metalama.Framework.Engine.Pipeline.ServiceProvider.^NKXN6diP.^Ib9O1__tTiGv(IServiceProvider _0)
   at Metalama.Framework.Engine.Pipeline.ServiceProvider.^j47JJ8dp.^OD2cX6Vs(IServiceProvider _0)
   at Metalama.Framework.Engine.Pipeline.ServiceProvider.^HFf83eaPqWlN(Type _0)
   at Metalama.Framework.Project.ServiceProviderExtensions.GetRequiredService[T](IServiceProvider serviceProvider)
   at Obfuscated.^Qf3QRcMSatsB..ctor(IServiceProvider serviceProvider)
   at Metalama.Framework.Engine.Pipeline.ServiceProvider.^LUKZAb4R(IEnumerable`1 _0, Compilation _1)
   at Metalama.Framework.DesignTime.VisualStudio.Classification.DesignTimeClassificationService..ctor()
   at Metalama.Framework.DesignTime.VisualStudio.CompilerServiceProvider.GetService[T]()
   at Metalama.Vsx.TextBuffers.TextBufferServicesImpl.<GetClassificationsImplAsync>d__18.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Metalama.Vsx.TextBuffers.TextBufferServicesImpl.<GetClassificationsAsync>d__17.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Metalama.Vsx.TextBuffers.TextBufferServicesImpl.<GetQuickInfoItemAsync_Source>d__26.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Metalama.Vsx.TextBuffers.TextBufferServicesImpl.<GetQuickInfoItemAsync>d__25.MoveNext()

When attempting to run Show Metalama Diff, I receive this error:
"Cannot preview the transformed code: This is not a Metalama project."

**Reply by gfraiteur** [reply]
@bradlee-s  Could you create a new conversation since it seems a different topic?
